include/audio_module/audio_module.h

```diff
@@ -0,0 +1,500 @@
+	bool use_tx_queue;
+
+	/* List node (pointer to next audio module). */
+	sys_snode_t node;
+
+	/* A single linked-list of the handles the module is connected to. */
+	sys_slist_t hdl_dest_list;
+
+	/* Number of destination modules. */
+	uint8_t dest_count;
+
+	/* Semaphore to count messages between modules. */
+	struct k_sem sem;
+
+	/* Mutex to make the above destinations list thread safe. */
+	struct k_mutex dest_mutex;
+
+	/* Module's thread configuration. */
+	struct audio_module_thread_configuration thread;
+
+	/* Private context for the module. */
+	struct audio_module_context *context;
+};
+
+/**
+ * @brief Private structure describing a data_in message into the module thread.
+ */
+struct audio_module_message {
+	/* Audio data to input. */
+	struct audio_data audio_data;
+
+	/* Sending module's handle. */
+	struct audio_module_handle *tx_handle;
+
+	/* Callback for when the audio data has been consumed. */
+	audio_module_response_cb response_cb;
+};
+
+/**
+ * @brief Open an audio module.
+ *
+ * @param parameters[in]     Pointer to the module set-up parameters.
+ * @param configuration[in]  Pointer to the module's configuration.
+ * @param name[in]           A NULL terminated string giving a unique name for this module instance.
+ * @param context[in/out]    Pointer to the private context for the module.
+ * @param handle[out]        Pointer to the module's private handle.
+ *
+ * @return 0 if successful, error otherwise.
+ */
+int audio_module_open(struct audio_module_parameters const *const parameters,
+		      struct audio_module_configuration const *const configuration,
+		      char const *const name, struct audio_module_context *context,
+		      struct audio_module_handle *handle);
+
+/**
+ * @brief Close an audio open module.
+ *
+ * @param handle[in/out]  The handle to the module instance.
+ *
+ * @return 0 if successful, error otherwise.
+ */
+int audio_module_close(struct audio_module_handle *handle);
+
+/**
+ * @brief Reconfigure an audio module.
+ *
+ * @param handle[in/out]     The handle to the module instance.
+ * @param configuration[in]  Pointer to the module's configuration to set.
+ *
+ * @return 0 if successful, error otherwise.
+ */
+int audio_module_reconfigure(struct audio_module_handle *handle,
+			     struct audio_module_configuration const *const configuration);
+
+/**
+ * @brief Get the configuration of an audio module.
+ *
+ * @param handle[in]          The handle to the module instance.
+ * @param configuration[out]  Pointer to the module's current configuration.
+ *
+ * @return 0 if successful, error otherwise.
+ */
+int audio_module_configuration_get(struct audio_module_handle const *const handle,
+				   struct audio_module_configuration *configuration);
+
+/**
+ * @brief Connect two audio modules together.
+ *
+ * @param handle_from[in/out]   The handle for the module for output.
+ * @param handle_to[in/out]     The handle of the module for input, should be NULL if
```

Comment:
Isn't the goal here that data can be exchanged between modules while also getting the data up to e.g. main.c?

Reply:
Same for audio_module_disconnect()

Reply:
No the module can send the output only back to the app, only onto the next module(s) or back to the app and onto the next module(s).